lib/signal-handler.c

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2018 Balabit
+ * Copyright (c) 2018 Kokan
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include "signal-handler.h"
+
+#include "syslog-ng.h"
+#include "children.h"
+
+#ifndef _WIN32
+#include <dlfcn.h>
+#endif
+
+static const struct sigaction *sgchld_handler = NULL;
+
+void
+trigger_sigchld_handler_chain(int signum)
+{
+  if (sgchld_handler && sgchld_handler->sa_handler)
+    {
+      sgchld_handler->sa_handler(signum);
+    }
+}
+
+#ifndef _WIN32
+static int
+call_original_sigaction(int signum, const struct sigaction *act, struct sigaction *oldact)
+{
+  static int (*real_sa)(int, const struct sigaction *, struct sigaction *) = NULL;
+
+  if (real_sa == NULL)
+    {
+      real_sa = dlsym(RTLD_NEXT, "sigaction");
```

Comment:
> Right now it is called signal chaining, but only one handler is saved and triggered. Is it okay to save just the latest and do not trigger the other handlers ?

Would the problem resolved by collection the function pointer into GPtrArray. It would not introduce much extra code complexity, considering how tricky the implementation is.

Reply:
We concluded with @Kokan that my suggestion wouldn't work. Other modules instances could be written with actual care, and save the original function pointer before overwriting. So in the end, I could not decide if a pointer in GPtrArray should be called or not.

I agree with the current version.